File: system-design/oop-patterns/01-strategy-pattern/full/js/main.js
import { Duck } from './duck/Duck.js';
import { SimpleFlyBehavior } from './duck/SimpleFlyBehavior.js';
import { JetFlyBehavior } from './duck/JetFlyBehavior.js';
import { NoFlyBehavior } from './duck/NoFlyBehavior.js';
import { SimpleQuackBehavior } from './duck/SimpleQuackBehavior.js';
import { NoQuackBehavior } from './duck/NoQuackBehavior.js';

const wildDuck = new Duck(new SimpleFlyBehavior(), new SimpleQuackBehavior());
const rubberDuck = new Duck(new NoFlyBehavior(), new NoQuackBehavior());
const mountainDuck = new Duck(new JetFlyBehavior(), new SimpleQuackBehavior());

wildDuck.performQuack();
wildDuck.performFly();

rubberDuck.performQuack();
rubberDuck.performFly();

mountainDuck.performQuack();
mountainDuck.performFly();

File: system-design/oop-patterns/01-strategy-pattern/full/java/Main.java
import duck.Duck;
import duck.SimpleFlyBehavior;
import duck.JetFlyBehavior;
import duck.NoFlyBehavior;
import duck.SimpleQuackBehavior;
import duck.NoQuackBehavior;

public class Main {
    public static void main(String[] args) {
        Duck wildDuck = new Duck(new SimpleFlyBehavior(), new SimpleQuackBehavior());
        Duck rubberDuck = new Duck(new NoFlyBehavior(), new NoQuackBehavior());
        Duck mountainDuck = new Duck(new JetFlyBehavior(), new SimpleQuackBehavior());

        wildDuck.performQuack();
        wildDuck.performFly();

        rubberDuck.performQuack();
        rubberDuck.performFly();

        mountainDuck.performQuack();
        mountainDuck.performFly();
    }
}
